Fix prop cycling not working

cvanelteren opened the pull request:
Addresses #25 

~~Note this is currently a draft as we are directly interfacing with private functions and this is not preferred.~~ But it does solve the issue. I will give it some thought on how this could be solved.

cvanelteren commented:
I would propose to change the cycle function to a cycle class and handle the cycling through there. This monkey patch here relies too much on internal mpl functions/variables and could be changed in the future.

cvanelteren commented:
Now the cycling is more robust, and passing the tests without checking with baseline compare. Will do it later.

cvanelteren commented:
Pinging @beckermr @pratiman-91

cvanelteren commented:
Could we go ahead and merge this? @pratiman-91 @beckermr

pratiman-91 commented:
I am out on holiday until the first week of Feb. If someone else can review. 
Thanks!

beckermr commented on ultraplot/axes/plot.py:
```diff
@@ -3482,14 +3479,17 @@ def _apply_scatter(self, xs, ys, ss, cc, *, vert=True, **kwargs):
             # Don't set 'c' explicitly unless it was provided
-            kw["s"] = s
-            if c is not None:
-                kw["c"] = c
+            kw["s"], kw["c"] = s, c
+            kw = self._parse_cycle(n, cycle_manually=cycle_manually, **kw)
```
This change doesn't match the comment. Why was it made? We need to either adjust the code or the comment. Also, why did you add another _parse_cycle call here?

cvanelteren replied:
This reverts the behavior of proplot.

cvanelteren replied:
See comment below

beckermr commented on ultraplot/axes/plot.py:
```diff
@@ -3482,14 +3479,17 @@ def _apply_scatter(self, xs, ys, ss, cc, *, vert=True, **kwargs):
                 infer_rgb=infer_rgb,
                 **kw,
             )
+        # Move _parse_cycle before _parse_color
+        kw = self._parse_cycle(
+            xs.shape[1] if xs.ndim > 1 else 1, cycle_manually=cycle_manually, **kw
+        )
```
The comment mentions moving the cycler parsing before the parsing of the color. However, this change undoes that. Why? Should we delete the comment or do we need to move this back?

beckermr commented on ultraplot/axes/plot.py:
```diff
@@ -3459,12 +3459,9 @@ def _apply_scatter(self, xs, ys, ss, cc, *, vert=True, **kwargs):
 
-        # Move _parse_cycle before _parse_color
-        kw = self._parse_cycle(xs.shape[1] if xs.ndim > 1 else 1, **kw)
-
         # Only parse color if explicitly provided
+        infer_rgb = True
         if cc is not None:
-            infer_rgb = True
```
This appears to be a change in defaults. Can you add a comment as to why or put it back?

cvanelteren replied:
I moved this back t o how proplot does it. I initially thought I monkey patched in cycling behavior which broke on the original PR I had with proplot. The colors were cycling but the other properties were not. Not is does all.

beckermr commented on ultraplot/constructor.py:
```diff
@@ -762,148 +762,187 @@ def _pop_modification(key):
+class Cycle(cycler.Cycler):
+    def __init__(self, *args, N=None, samples=None, name=None, **kwargs):
+        """
+        Generate and merge `~cycler.Cycler` instances in a variety of ways.
```
The docs here no longer match the fact that this is a class as opposed to a function. We should adjust.

cvanelteren replied:
Good catch. I just copied the doc string over -- will fix.

beckermr commented on ultraplot/constructor.py:
```diff
@@ -762,148 +762,187 @@ def _pop_modification(key):
+        Returns
+        -------
+        cycler.Cycler
+            A `~cycler.Cycler` instance that can be passed
```
This is no longer true. We now have a new class.

beckermr reviewed (changes requested):
Verifying that the Cycler object is correctly instantiated by our sublcass is difficult. I'd be good to add specific unit tests of this so that we verify it is correct and we can catch regressions. Ideally, we'd write the tests against a working version of the factory function first and then move the change to a class to a new PR and ensure the tests are still passing.

cvanelteren commented:
> Verifying that the Cycler object is correctly instantiated by our sublcass is difficult. I'd be good to add specific unit tests of this so that we verify it is correct and we can catch regressions. Ideally, we'd write the tests against a working version of the factory function first and then move the change to a class to a new PR and ensure the tests are still passing.

Using #20, we can check for validity. That is, as soon as we have main producing a good baseline, any further tests can be used to check whether changes to the codebase will produce the same validity as we will expect (using pytest-mpl). Currently main does not pass these tests, but I am expecting that the joint of #26, #24 will.

cvanelteren commented:
I do like the idea of just testing if cycling happens correctly. I will write a simple test to see if it does that is lower level than the high level tests proposed above.

cvanelteren commented:
> I am out on holiday until the first week of Feb. If someone else can review. 
> Thanks!

Have fun😬

cvanelteren reviewed (commented):
Applied all fixed, added unittests, fixed up doc strings

beckermr commented on ultraplot/constructor.py:
```diff
@@ -762,148 +773,182 @@ def _pop_modification(key):
-    """
-    Generate and merge `~cycler.Cycler` instances in a variety of ways.
+class Cycle(cycler.Cycler):
+    def __init__(self, *args, N=None, samples=None, name=None, **kwargs):
```
Doc strings for classes are always added to the class, not the __init__ function.

beckermr reviewed (changes requested):
One more item.

cvanelteren commented:
![image](https://github.com/user-attachments/assets/af625876-c54d-44b9-a012-3392ff418b6d)
putting it back to draft -- I have been scratching my head why this is not working. Will keep this posted until I can fix this. The cycler properly generates a map from the keyword cycler but it is not properly passed on to pie.

cvanelteren commented:
Rebased to changes to clean up the mess -- should be working now. I cleaned up the logic inside `_parse_cycle` as I found it difficult to follow exactly what was returned where and when. This way I can more easily follow. As far as I can tell the cycling is working properly now. However, I noticed one crucial fact. UltraPlot performs a step at draw that may change the look of the images. This is not picked up when the image is passed to pytest-mpl.  I will run some further tests tomorrow and then this is done.

I'll put it on the todo to check what is causing this.

cvanelteren commented:
I am not very with how I cycle the properties here: there are calls to private properties and I cycle the object manually. After digging a bit, I can see that there is  currently no public api for accessing the cycling object (see https://github.com/matplotlib/matplotlib/issues/26831#issuecomment-1781552042). This leads us at a crossroad

1. We can either provide a public api
2. Or cycle it manually 

UltraPlot does override some public function that exist already (e.g. see `set_prop_cycle`).

I am also not sure if this goes beyond this PR -- intended as a quick fix -- rather than a larger api change. Open for suggestions.

cvanelteren commented:
Also see https://github.com/matplotlib/matplotlib/pull/29469
Which makes me lean towards merging it as is and waiting for potential future api to change such that we can follow suit.

cvanelteren commented:
Then we could add a test this future PR with

```python

def test_manual_cycle():
    cycle = uplt.Cycle(
        ["red", "green", "black"],
        marker=["X", "o"],
        sizes=[20, 100],
        edgecolors=["r", "k"],
    )
    fig, ax = uplt.subplots()
    ax.set_prop_cycle(cycle)
    ax.scatter([1], [1])

    ax = ax[0]  # force to get single axis rather than GridSpec
    parser = ax._get_lines
    current_pos = parser._idx
    cycled_pos = (current_pos + 1) % len(parser._cycler_items)
    props = ax._active_cycle.get_next()
    assert cycled_pos == parser._idx
    assert props == parser._cycler_items[parser._idx]
```

or something similar

cvanelteren commented:
I belief the root cause of the baseline images not showing the proper formatting when using pytest-mpl and they look fine when manually plotting is that in pytest-mpl they remove the `rcContex` prior to plotting: https://github.com/matplotlib/pytest-mpl/blob/29fd89095ced454e1f82007aeb592268a430aba0/pytest_mpl/plugin.py#L730

cvanelteren commented:
I am merging this with devel. After locally testing the open PRs it all looks good for now.

cvanelteren commented:
is that ok @beckermr?

beckermr commented on ultraplot/axes/plot.py:
```diff
@@ -2477,54 +2477,43 @@ def _parse_cycle(
-        else:
-            return kwargs
+            return resolved_cycle, kwargs
+
+        if resolved_cycle and resolved_cycle != self._active_cycle:
+            self.set_prop_cycle(resolved_cycle)
+
+        # Apply manual cycle properties
+        if cycle_manually:
+            current_prop = self._get_lines._cycler_items[self._get_lines._idx]
+            self._get_lines._idx = (self._get_lines._idx + 1) % len(self._active_cycle)
+            for prop, key in cycle_manually.items():
+                if kwargs.get(key) is None and prop in current_prop:
+                    value = current_prop[prop]
+                    kwargs[key] = pcolors.to_rgba(value) if key == "c" else value
```
With the new changes, this block of code is skipped if `return_cycle` is `True`. Why is that?

beckermr commented on ultraplot/constructor.py:
```diff
@@ -845,65 +850,103 @@ def Cycle(*args, N=None, samples=None, name=None, **kwargs):
+        maxlen = np.lcm.reduce(lengths)
+        props = {key: value * (maxlen // len(value)) for key, value in props.items()}
+        mcycler = cycler.cycler(**props)
+        super().__init__(mcycler)
```
Can you add a comment on what this call does exactly? Does it use matplotlib's init function to update the current cycle to the computed one?

cvanelteren replied:
`cycler.cycler` is not a class but a function in mpl. The logic of the init from the `Cycle` class is not compatible with the function -- that is we need to pass the output of cycler directly to the parent of the class we introduced.

beckermr replied:
Cool. Let's add a comment on this in the code.

beckermr commented on ultraplot/constructor.py:
```diff
@@ -845,65 +850,103 @@ def Cycle(*args, N=None, samples=None, name=None, **kwargs):
-    if not args:
-        props.setdefault("color", ["black"])
+    def __init__(self, *args, N=None, samples=None, name=None, **kwargs):
+        self.name = "_no_name"  # default value
```
This is a change in behavior. In the previous code, the cycler always got the input name unless the function hit the early return. I'd use the `_not_none` here too.

cvanelteren replied:
Right again!

beckermr commented on ultraplot/axes/plot.py:
```diff
@@ -2477,54 +2477,43 @@ def _parse_cycle(
+            for prop, key in cycle_manually.items():
+                if kwargs.get(key) is None and prop in current_prop:
+                    value = current_prop[prop]
+                    kwargs[key] = pcolors.to_rgba(value) if key == "c" else value
```
This is to_hex previously. Why the change?

cvanelteren replied:
You are right! Shouldn't have happened. I believe I was testing something here and forgot to change it back. See #29

beckermr reviewed (changes requested):
This PR needs changes. Please wait next time.

cvanelteren commented:
Ah my bad!

cvanelteren commented:
I will open another PR to fix the fixes.

beckermr commented:
Thank you! Sorry for the delay in my code review.

cvanelteren commented:
No worries ;-)
